src/NexusMods.App.UI/WorkspaceSystem/Controllers/WorkspaceController.cs

```diff
@@ -137,32 +161,87 @@ public void ChangeActiveWorkspace(WorkspaceId workspaceId)
         }
     }
 
+    /// <inheritdoc/>
+    public IWorkspaceViewModel ChangeOrCreateWorkspaceByContext<TContext>(Func<Optional<PageData>> getPageData) where TContext : IWorkspaceContext, new()
+    {
+        Dispatcher.UIThread.VerifyAccess();
+
+        if (!TryGetWorkspaceByContext<TContext>(out var existingWorkspace))
+        {
+            var pageData = getPageData();
+
+            var newWorkspace = CreateWorkspace(
+                new TContext(),
+                pageData
+            );
+
+            existingWorkspace = newWorkspace;
+        }
+
+        ChangeActiveWorkspace(existingWorkspace.Id);
+        return existingWorkspace;
+    }
+
+    /// <inheritdoc/>
+    public IWorkspaceViewModel ChangeOrCreateWorkspaceByContext<TContext>(Func<TContext, bool> predicate, Func<Optional<PageData>> getPageData,
+        Func<TContext> getWorkspaceContext) where TContext : IWorkspaceContext
+    {
+        Dispatcher.UIThread.VerifyAccess();
+
+        var workspaces = FindWorkspacesByContext<TContext>();
+        var existingWorkspace = workspaces.FirstOrOptional(tuple => predicate(tuple.Item2));
+
+        if (!existingWorkspace.HasValue)
+        {
+            var newWorkspace = CreateWorkspace(
+                getWorkspaceContext(),
+                getPageData()
+            );
+
+            ChangeActiveWorkspace(newWorkspace.Id);
+            return newWorkspace;
+        }
+
+        ChangeActiveWorkspace(existingWorkspace.Value.Item1.Id);
+        return existingWorkspace.Value.Item1;
+    }
+
     public void AddPanel(WorkspaceId workspaceId, WorkspaceGridState newWorkspaceState, AddPanelBehavior behavior)
     {
+        Dispatcher.UIThread.VerifyAccess();
```

Comment:
Is this temporary for debug or is that going to be in the final code?
I would rather not have it

Reply:
The controller can only work on the UI thread. Using it in another thread is not allowed.